controllers/provider-azure/pkg/apis/azure/helper/scheme.go

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package helper
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/gardener/gardener-extensions/pkg/controller"
+	"github.com/gardener/gardener-extensions/pkg/util"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+
+	api "github.com/gardener/gardener-extensions/controllers/provider-azure/pkg/apis/azure"
+	"github.com/gardener/gardener-extensions/controllers/provider-azure/pkg/apis/azure/install"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+)
+
+var (
+	// Scheme is a scheme with the types relevant for OpenStack actuators.
+	Scheme *runtime.Scheme
+
+	decoder runtime.Decoder
+)
+
+func init() {
+	Scheme = runtime.NewScheme()
+	utilruntime.Must(install.AddToScheme(Scheme))
+
+	decoder = serializer.NewCodecFactory(Scheme).UniversalDecoder()
+}
+
+// InfrastructureConfigFromInfrastructure extracts the InfrastructureConfig from the
+// ProviderConfig section of the given Infrastructure.
+func InfrastructureConfigFromInfrastructure(infra *extensionsv1alpha1.Infrastructure) (*api.InfrastructureConfig, error) {
+	config := &api.InfrastructureConfig{}
+	if infra.Spec.ProviderConfig != nil && infra.Spec.ProviderConfig.Raw != nil {
+		if _, _, err := decoder.Decode(infra.Spec.ProviderConfig.Raw, nil, config); err != nil {
+			return nil, err
+		}
+
+		return config, nil
+	}
+	return nil, fmt.Errorf("infrastructure config is not set on the infrastructure resource")
```

Comment:
I do not understand this error message.

Reply:
It is the original method, I now changed it to `provider config`.